File: boj/galaxy4276/4949.java
import java.util.*;
import java.io.*;

class Solution {

  public boolean isBalance(Stack<Character> stack, String str) {
    for (char c : str.toCharArray()) {
      if (c == '[' || c == '(') stack.add(c);
      else if (c == ')') {
        if (stack.isEmpty() || stack.peek() != '(') return false;
        stack.pop();
      } else if (c == ']') {
        if (stack.isEmpty() || stack.peek() != '[') return false;
        stack.pop();
      }
    }
    return stack.isEmpty();
  }

}

public class Main {
  public static void main(String[] args) throws IOException {
    final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    final BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));
    final Solution s = new Solution();

    while (true) {
      String readed = br.readLine();
      if (readed.equals(".")) break;
      var stack = new Stack<Character>();
      boolean isBalance = s.isBalance(stack, readed);
      if (isBalance) writer.write("yes\n");
      else writer.write("no\n");
    }

    writer.flush(); writer.close();
  }
}

File: boj/pandora2948/silver/silver4/4949.py
from sys import stdin

while True:
    msg = stdin.readline()
    if msg == '.\n':
        break
    stack = list()
    integrity = True
    try:
        for char in msg:
            if char == '.':
                break

            if char == '[':
                stack.append('[')

            elif char == ']':
                if stack[-1] == '[':
                    stack.pop()

                else:
                    integrity = False
                    break

            elif char == '(':
                stack.append('(')

            elif char == ')':
                if stack[-1] == '(':
                    stack.pop()

                else:
                    integrity = False
                    break

        if not stack and integrity:
            print('yes')

        else:
            print('no')

    except IndexError:
        print('no')

